### ISSUE TITLE
Debug interactive closes and debugging stops when there is an error

### ISSUE THREAD

**Issue by bschnurr** [microsoft/PTVS#5827]
@bschnurr commented on [Mon Oct 07 2019](https://github.com/microsoft/ptvsd/issues/1827)

@jose-alex commented on [Sat Oct 05 2019](https://github.com/microsoft/PTVS/issues/5800)

In VS 2019, say I am debugging a project and the debugger stops at a breakpoint. If I write a command in the Debug Interactive window and it has an error (syntax or otherwise) the window closes and debugging stops. This was not so in VS 2017 and I think it is not a desirable behavior. Any small error like the one in the image will do it.
![imagen](https://user-images.githubusercontent.com/6751250/66264986-1731e000-e7cc-11e9-9d78-547fc7c0234b.png)
Is there a way to configure VS 2019 to avoid this behavior?

---

@bschnurr commented on [Mon Oct 07 2019](https://github.com/microsoft/PTVS/issues/5800#issuecomment-539137851)

Added debugger logging below
https://docs.microsoft.com/en-us/visualstudio/python/debugging-python-in-visual-studio?view=vs-2019#enable-debugger-logging

```
 1> [DebugAdapter] --> C (stackTrace-25): {"type":"request","command":"stackTrace","arguments":{"threadId":1},"seq":25}
 1> [DebugAdapter] <--   R (stackTrace-25) [2 ms]: {"type": "response", "seq": 32, "request_seq": 25, "success": true, "command": "stackTrace", "message": "", "body": {"stackFrames": [{"id": 2, "name": "<module>", "line": 3, "column": 1, "source": {"path": "C:\\Users\\xxx\\source\\repos\\PythonApplication30\\PythonApplication30.py", "sourceReference": 0}}], "totalFrames": 1}}
 1> [DebugAdapter] --> C (evaluate-26): {"type":"request","command":"evaluate","arguments":{"context":"repl","expression":"da\r@LINE@","frameId":2},"seq":26}
 1> [DebugAdapter] <--   R (evaluate-26) [21 ms]: {"type": "response", "seq": 33, "request_seq": 26, "success": false, "command": "evaluate", "message": "NameError: name 'da' is not defined\n", "body": {}}
 1> ERROR: Unexpected error

ProtocolException: NameError: name 'da' is not defined

Microsoft.VisualStudio.Shared.VSCodeDebugProtocol.Protocol.DebugProtocol.SendRequestSyncCore[TArgs,TResponse](IDebugRequestWithResponse`2 request)
Microsoft.VisualStudio.Shared.VSCodeDebugProtocol.DebugProtocolHost.SendRequestSync[TArgs,TResponse](DebugRequestWithResponse`2 request)
Microsoft.VisualStudio.Debugger.VSCodeDebuggerHost.Engine.Implementation.CustomProtocolManager.ProtocolHostWrapper.Microsoft.VisualStudio.Debugger.DebugAdapterHost.Interfaces.IProtocolHostOperations.SendRequestSync[TArgs,TResponse](DebugRequestWithResponse`2 request)

 1> ERROR: NameError: name 'da' is not defined
 1> Debug adapter process exited.

```

---

@karthiknadig commented on [Mon Oct 07 2019](https://github.com/microsoft/ptvsd/issues/1827#issuecomment-539162412)

@bschnurr Are you seeing a difference in the text that is sent to the debugger? between 2017 and 2019? 

---

@bschnurr commented on [Mon Oct 07 2019](https://github.com/microsoft/ptvsd/issues/1827#issuecomment-539181806)

legacy debugger sends error text to the debug interactive window. 
new debugger shows a popup and disconnects

---

@karthiknadig commented on [Mon Oct 07 2019](https://github.com/microsoft/ptvsd/issues/1827#issuecomment-539273242)

@bschnurr I was asking for the logs like above from 2017 for the new debugger.

---

@fabioz commented on [Tue Oct 08 2019](https://github.com/microsoft/ptvsd/issues/1827#issuecomment-539640613)

I'm investigating this one... 

I was able to reproduce this locally (the related log files are below).

[ptvsd-9812.log](https://github.com/microsoft/ptvsd/files/3703972/ptvsd-9812.log)
[pydevd.log](https://github.com/microsoft/ptvsd/files/3703973/pydevd.log)

---

@fabioz commented on [Tue Oct 08 2019](https://github.com/microsoft/ptvsd/issues/1827#issuecomment-539654344)

So, after investigating it seems that if the evaluate response has `success=False` visual studio kills the program (i.e.: by changing `success` to `True` in https://github.com/microsoft/ptvsd/blob/master/src/ptvsd/_vendored/pydevd/_pydevd_bundle/pydevd_comm.py#L921 everything works).

Now, I'm not exactly sure how to proceed here. The choices I can think of are:
- Do a workaround to return `success=True` if the client is visual studio in the evaluate request.
- Always return `True` in the evaluate request (for any client) regardless of whether the evaluation result was actually gotten from an exception trying to evaluate the expression (`debugProtocol.json` doesn't really specify what `success=False` means in the evaluate context... having `success=False` does seem correct to me in this case, but maybe I'm not interpreting it properly -- VSCode does work properly though).
- Pass the issue back to the Visual Studio team for them to fix it on their side. It doesn't seem like it should stop the debugging in this case -- there are a few other cases where `success=False` is expected and I don't expect the debugging to finish because of that. If that's the case we have to re-evaluate those other use-cases too (we should clarify with the Visual Studio team if `success=False` always stops debugging or if this is an isolated case/bug).

@karthiknadig @int19h what do you think?

@bschnurr you can work around locally by changing `'success':False` to `'success':True` locally (the file to be changed is `C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\Common7\IDE\Extensions\Microsoft\Python\Core\Packages\ptvsd\_vendored\pydevd\_pydevd_bundle\pydevd_comm.py` -- the occurrence is inside the `def _evaluate_response` method).

---

@bschnurr commented on [Tue Oct 08 2019](https://github.com/microsoft/ptvsd/issues/1827#issuecomment-539659733)

I'll try that. I see the same ```success=False``` in the older 4.1.1a9 version of ptvsd

[ptvsd_4.1.1a9.log](https://github.com/microsoft/ptvsd/files/3704133/ptvsd_4.1.1a9.log)

---

@fabioz commented on [Tue Oct 08 2019](https://github.com/microsoft/ptvsd/issues/1827#issuecomment-539662782)

@bschnurr I didn't realize you were part of PTVS, so, great, it seems the PTVS team is already on the loop ;)

Please let me know if it fixes for you (and we should still see how to address the issue properly).

---

@fabioz commented on [Tue Oct 08 2019](https://github.com/microsoft/ptvsd/issues/1827#issuecomment-539663330)

> I'll try that. I see the same `success=False` in the older 4.1.1a9 version of ptvsd
> 
> [ptvsd_4.1.1a9.log](https://github.com/microsoft/ptvsd/files/3704133/ptvsd_4.1.1a9.log)

Actually, in that log it seems that `"success": true, "command": "evaluate"`.

---

@bschnurr commented on [Tue Oct 08 2019](https://github.com/microsoft/ptvsd/issues/1827#issuecomment-539664796)

yes that seems to be the old behaviour of ptvsd

---

@darindillon commented on [Sun Oct 13 2019](https://github.com/microsoft/ptvsd/issues/1827#issuecomment-541482224)

Additional info -- the problem is not VS 2017 vs VS 2019. Enabling the "Use legacy debugger" in VS 2019 (Tools->Options->Python->Debugging->"Use legacy debugger") completely fixes this issue and I can now type a syntax error in the debugger without VS shutting down the entire program.

**Comment by huguesv** [comment]
See the email thread "Evaluate request handling" for more info.

Summary here:

You can control your own error handling.  I would suggest modifying EvaluateReplRequest [0] (and any other methods) so that it calls SendRequest instead of SendRequestSync – this lets you register an error handling callback.  SendRequest just tears down the session on failure, and it would be a potentially breaking API change to change this.

I also noticed that you're maintaining a static instance of your custom protocol extension.  This is likely to cause problems when debugging multiple processes, as VS runs one instance of the debugger per process, and each debugger instance will have its own instances of your PZ extensibility objects.  

We made a change a while back to allow engine registrations to supply a list of generic "ExtensibilityObjects", rather than requiring the use of separate properties like the existing "AdapterLauncher" and "CustomProtocolExtension".  With this, you could have a single component that implements both IAdapterLauncher and ICustomProtocolExtension, which will be instantiated only once per debugger instance, avoiding the need to coordinate between multiple components.

Basically, you'd change your pkgdef from something like:
```
[$RootKey$\AD7Metrics\Engine\{engine-guid}]
; ...
"AdapterLauncher"="{adapter-launcher-guid}"
"CustomProtocolExtension"="{custom-protocol-guid}"
; ...
```
To:
```
[$RootKey$\AD7Metrics\Engine\{engine-guid}]
; ...
; Remove "AdapterLauncher" and "CustomProtocolExtension" properties
; ...

[$RootKey$\AD7Metrics\Engine\{engine-guid}\ExtensibilityObjects]
"1"="{combined-launcher-and-protocol-extension-guid}"

```

**Comment by huguesv** [comment]
The attribute that generates this pkgdef is probably `ProvideDebugAdapterAttribute`

**Comment by innominate227** [comment]
I still see this issue in VisualStudio 16.4.2, it made make debugging python code very painful.  Is this still not released?

**Comment by huguesv** [comment]
@innominate227 This went in too late to make it into 16.4. I'm pretty sure it's released in 16.5 Preview 1 (I can't verify right now, I only have 16.5 Preview 2 builds installed)
